script to generate metabrlists from data coming from my metabarcoding processing pipeline (OBIholo_pipeline).

```{r}
#  remotes::install_github("metabaRfactory/metabaR")
library(metabaR)
#  remotes::install_github("LafontRapnouilTristan/EcophyCofog")
EcophyCofog::Library(c("readr","magrittr","dplyr","stringr","reshape2",
                       "ggplot2","ggpubr","forcats","cowplot","phyloseq",
                       "speedyseq"))
```

# Input format

The basic data format used in `metabaR` is a `metabarlist`, a list of four tables:   

- `reads` a table of class `matrix` consisting of PCRs as rows, and molecular operational taxonomic units (MOTUs) as columns. The number of reads for each MOTU in each PCR is given in each cell, with 0 corresponding to no reads.   

- `motus` a table of class `data.frame` where MOTUs are listed as rows, and their attributes as columns. **A mandatory field in this table is the field "sequence", i.e. the DNA sequence representative of the MOTU**.  Examples of other attributes that can be included in this table are the MOTU taxonomic information, the taxonomic assignment scores, MOTU sequence GC content, MOTU total abundance in the dataset, etc. 

- `pcrs` a table of class `data.frame` consisting of PCRs as rows, and PCR attributes as columns. This table is particularly important in `metabaR`, as it contains all the information related to the extraction, PCR and sequencing protocols and design that are necessary to assess and improve the quality of metabarcoding data [@taberlet2018environmental; @zinger2019dna]. This table can also include information relating to the PCR design, such as the well coordinates and plate of each PCR, the tag combinations specific of the PCR, the primers used, etc. **Mandatory fields are**:    
    - **`sample_id`**: a vector indicating the biological sample origin of each PCR (e.g. the sample name)
    - **`type`** : the type of PCR, either a sample or an experimental control amplification. Only two values allowed: `"sample"` or `"control"`.     
    - **`control_type`** : the type of control. Only five values are possible in this field:    
        - `NA` if `type="sample"`, i.e. for any PCR obtained from a biological sample.    
        - `"extraction"` for DNA extraction negative controls, i.e. PCR amplification of an extraction where the DNA template was replaced by extraction buffer.   
        - `"pcr"` for PCR negative controls, i.e. pcr amplification where the DNA template was replaced by PCR buffer or sterile water.    
        - `"sequencing"` for sequencing negative controls, i.e. unused tag/library index combinations.     
        - `"positive"` for DNA extraction or PCR positive controls, i.e. pcr amplifications of known biological samples or DNA template (e.g. a mock community).     

- `samples` a table of class `data.frame` consisting of biological samples as rows, and associated information as columns. Such information includes e.g. geographic coordinates, abiotic parameters, experimental treatment, etc. This table does not include information on the DNA metabarcoding experimental controls, which can only be found in `pcrs`.   

# Prep files

Load required files:

```{r}
folder_names <- list.files("data/OBI_pipeline_results/")
folder_names <- folder_names[c(1:2)]

list_motus <- list()
list_reads <- list()
list_filters <- list()

for (i in 1:length(folder_names)){
  current_name <- folder_names[i] 
  
  list_motus[[current_name]] <- read.csv2(paste0("data/OBI_pipeline_results/",current_name,"/",current_name,"_taxassigned.csv"))
  
  list_reads[[current_name]] <- read_delim(paste0("data/OBI_pipeline_results/",current_name,"/",current_name,"_R1R2_good_demultiplexed_basicfilt_derepl_cl_agg.tab"),"\t", escape_double = F, trim_ws = T)
  
  list_filters[[current_name]] <- read_delim(paste0("data/OBI_pipeline_ngs/",current_name,"_ngsfilter.tab"),"\t", escape_double = F, trim_ws = T, col_names = F)
} 

list_filters$jm_ITS2 %<>% filter(X1!="INTER_Matthieu-0")
list_filters$jm_ITS2_2 %<>% filter(X1!="INTER_Matthieu-0")
```

get MOTUS counts in data_motus
```{r}
for (i in 1:length(folder_names)){
  
  list_motus[[i]] <- list_motus[[i]][match(toupper(list_reads[[i]]$sequence ),list_motus[[i]]$sequences ),]
  list_motus[[i]]$counts <- list_reads[[i]]$count
  
} 
```

prepare data_reads
```{r}
list_motus_seq <- list()
for (i in 1:length(list_reads)){
  
  list_motus_seq[[folder_names[i]]] <-  list_reads[[i]][,c("id","sequence")] 
  sample_col <- which(grepl("sample:",colnames(list_reads[[i]]))==T)
  tmp <- list_reads[[i]][,c(1,sample_col)]
  tmp1 <- t(tmp[,-1])
  id <- t(tmp[,1])
  colnames(tmp1) <- id
  rownames(tmp1) <- gsub("sample:","",gsub("-","_",rownames(tmp1)))
  list_reads[[i]] <- tmp1 
  
} 

list_reads[[1]] <-   as.data.frame(list_reads[[1]]) %>% slice(which(!grepl("_0",rownames(list_reads[[1]]))))
list_reads[[2]] <-   as.data.frame(list_reads[[2]]) %>% slice(which(!grepl("_0",rownames(list_reads[[2]]))))
```

prepare data_samples
```{r}
list_samples <- list()
for (i in 1:length(list_reads)){
  
  list_samples[[i]] <- data.frame(sample_id=rownames(list_reads[[i]]))
  
} 
```

prepare data_pcrs
```{r}
filtid <- list_filters$jm_ITS2$X2 # ID 16s
sampid <- list_samples[[1]]$sample_id # ID samples

list_filters$jm_ITS2$X2[which(filtid %in% sampid==F)] <- gsub("-","_",list_filters$jm_ITS2$X2[which(filtid %in% sampid==F)]) # replace - by _

filtid <- list_filters$jm_ITS2_2$X2 # ID 16s
sampid <- list_samples[[2]]$sample_id # ID samples
list_filters$jm_ITS2_2$X2[which(filtid %in% sampid==F)] <- gsub("-","_",list_filters$jm_ITS2_2$X2[which(filtid %in% sampid==F)])


c(setdiff(sort(list_filters[[1]]$X2),sort(list_samples[[1]]$sample_id))) # check if ngs and sample names match
c(setdiff(sort(list_samples[[1]]$sample_id),sort(list_filters[[1]]$X2)))

c(setdiff(sort(list_filters[[2]]$X2),sort(list_samples[[2]]$sample_id))) # check if ngs and sample names match
c(setdiff(sort(list_samples[[2]]$sample_id),sort(list_filters[[2]]$X2)))

list_filters$jm_ITS2 %<>% filter(X2!="CTED208A")
list_filters$jm_ITS2_2 %<>% filter(X2!="CTED208A")


list_pcrs <- list()
for (i in 1:length(list_reads)){
  
  colnames(list_filters[[i]]) <- c("project","ID","Tags","PrimeF","PrimeR","diff_doublons","x","Type") # rename columns
  
  # list_filters[[i]] <- dplyr::filter(list_filters[[i]], !ID %in% c(setdiff(sort(list_filters[[i]]$ID),sort(list_samples[[i]]$sample_id)))) #useless now
  
  list_pcrs[[i]] <- arrange(list_samples[[i]],sample_id) %>% mutate(type=ifelse(grepl("=control",arrange(list_filters[[i]],ID)$Type)==T,"control","sample"),
                                                                    control_type=ifelse(grepl("control_type=",arrange(list_filters[[i]],ID)$Type)==T,
                                                                                        str_extract(arrange(list_filters[[i]],ID)$Type,"(?<=control_type=)\\w+"),NA),
                                                                    plate= str_extract(arrange(list_filters[[i]],ID)$Type,"(?<=position=)[[:alnum:]]*"),
                                                                    well= str_extract(arrange(list_filters[[i]],ID)$Type,"(?<=_)[[:upper:]]{1}[[:digit:]]{1,2}"))
  
}
```

check names issues
```{r}
for (i in 1:length(list_reads)){
  print(setdiff(list_samples[[i]]$sample_id,list_filters[[i]]$ID))
  print(setdiff(list_filters[[i]]$ID,list_samples[[i]]$sample_id))
} 
```

finalize
```{r}
for (i in 1:length(list_reads)){
  
  rownames(list_motus[[i]]) <- sprintf("MOTU_%05d", 1:nrow(list_motus[[i]]))
  colnames(list_reads[[i]]) <- sprintf("MOTU_%05d", 1:ncol(list_reads[[i]]))
  
  list_pcrs[[i]] %<>% arrange(sample_id)
  list_filters[[i]] %<>% arrange(ID)
  
  rownames(list_pcrs[[i]]) <- list_pcrs[[i]]$sample_id
  colnames(list_motus[[i]])[1] <- "sequence" 
  
  list_pcrs[[i]] %<>% mutate(tag_fwd = str_extract(list_filters[[i]]$Tags,"^\\w+"), # get fwd tag seq
                             tag_rev = str_extract(list_filters[[i]]$Tags,"\\w+$"), # get rev tag seq
                             primer_fwd = list_filters[[i]]$PrimeF,
                             primer_rev = list_filters[[i]]$PrimeR,
                             plate_col = as.numeric(str_extract(well,"[:digit:]{1,2}")),# numeric fÃ¼r metabaR 
                             plate_row = str_extract(well,"[:upper:]"),
                             plate= as.numeric(str_extract(plate,"[:digit:]{1,2}")),
                             project = list_filters[[i]]$project
  ) %>%
    rename(plate_no = plate) 
  
  list_pcrs[[i]] %<>%
    mutate(control_type = ifelse(control_type=="ctab","extraction",control_type))%>%
    mutate(control_type = as.factor(control_type))
  
  rownames(list_samples[[i]]) <- list_samples[[i]]$sample_id
  
}
```

# Add sample biological information


Add info on my samples, has to be improved but i'll see when i'll get my data.
```{r}
list_samples[[1]] %<>%
 mutate(organ = ifelse( grepl("_R|(-R)|([0-9]{2}R)",list_samples[[1]]$sample_id)==T,"ROOTS",
                         ifelse( grepl("FL",list_samples[[1]]$sample_id)==T,"FLOWER",
                                 ifelse( grepl("FE",list_samples[[1]]$sample_id)==T,"LEAVES",
                                         ifelse( grepl("FR",list_samples[[1]]$sample_id)==T,"FRUIT",
                                                 ifelse( grepl("RJ",list_samples[[1]]$sample_id)==T,"RESHOOT",
                                                         ifelse( grepl("GR",list_samples[[1]]$sample_id)==T,"SEEDS",
                                                                 ifelse( grepl("ED",list_samples[[1]]$sample_id)==T,"ENDOPHYTES",
                                                                         ifelse( grepl("EP",list_samples[[1]]$sample_id)==T,"EPIPHYTES",
                                                                                 ifelse( grepl("[0-9]{2}(S|s|C)",list_samples[[1]]$sample_id)==T,"RHIZOSPHERE",
                                                                                         ifelse( grepl("TX|CT|EXT|PCR|TPOS|TEP|\\+",list_samples[[1]]$sample_id)==T,"CONTROL","SOIL")
                                                                                 )))))))))) %>%
  mutate(project = ifelse(organ %in% c("ENDOPHYTES","EPIPHYTES","RHIZOSPHERE") | grepl("[0-9]{2}R",list_samples[[1]]$sample_id)==T,"INT_M",list_pcrs[[1]]$project))


list_samples[[2]] %<>%
 mutate(organ = ifelse( grepl("_R|(-R)|([0-9]{2}R)",list_samples[[2]]$sample_id)==T,"ROOTS",
                         ifelse( grepl("FL",list_samples[[2]]$sample_id)==T,"FLOWER",
                                 ifelse( grepl("FE",list_samples[[2]]$sample_id)==T,"LEAVES",
                                         ifelse( grepl("FR",list_samples[[2]]$sample_id)==T,"FRUIT",
                                                 ifelse( grepl("RJ",list_samples[[2]]$sample_id)==T,"RESHOOT",
                                                         ifelse( grepl("GR",list_samples[[2]]$sample_id)==T,"SEEDS",
                                                                 ifelse( grepl("ED",list_samples[[2]]$sample_id)==T,"ENDOPHYTES",
                                                                         ifelse( grepl("EP",list_samples[[2]]$sample_id)==T,"EPIPHYTES",
                                                                                 ifelse( grepl("[0-9]{2}(S|s|C)",list_samples[[2]]$sample_id)==T,"RHIZOSPHERE",
                                                                                         ifelse( grepl("TX|CT|EXT|PCR|TPOS|TEP|\\+",list_samples[[2]]$sample_id)==T,"CONTROL","SOIL")
                                                                                 )))))))))) %>%
  mutate(project = ifelse(organ %in% c("ENDOPHYTES","EPIPHYTES","RHIZOSPHERE") | grepl("[0-9]{2}R",list_samples[[2]]$sample_id)==T,"INT_M",list_pcrs[[2]]$project))
```

# Generate metabarlists

```{r}
metablists <- NULL
list_pcrs[[1]] <- list_pcrs[[1]][,-c(4,10,11)]
list_pcrs[[2]] <- list_pcrs[[2]][,-c(4,10,11)]
for (i in 1:length(folder_names)){
  
  metablists[[folder_names[i]]] <- metabaR::metabarlist_generator(reads=as.matrix(list_reads[[i]]),
                                                               motus=list_motus[[i]],
                                                               pcrs=list_pcrs[[i]],
                                                               samples=list_samples[[i]])
}
```

Check if and which list have empty pcrs and motus.
```{r}
for (i in 1:length(folder_names)){
  data_temp <- metablists[[folder_names[i]]]
  if(sum(colSums(data_temp$reads)==0)>0){print(paste0(folder_names[i],": empty motus present"))}
  if(sum(rowSums(data_temp$reads)==0)>0){print(paste0(folder_names[i],": empty pcrs present"))}
}
```

Update
```{r}
for (i in 1:length(folder_names)){
  data_temp <- metablists[[folder_names[i]]]
  data_temp$motus$counts = colSums(data_temp$reads)
  data_temp$pcrs$nb_reads_postmetabaR = rowSums(data_temp$reads)
  data_temp$pcrs$nb_motus_postmetabaR = rowSums(ifelse(data_temp$reads>0, T, F))
  data_temp <- subset_metabarlist(data_temp,"motus",data_temp$motus$counts!=0)
  metablists[[folder_names[i]]] <- data_temp
}



for (i in 1:length(folder_names)){
  data_temp <- metablists[[folder_names[i]]]
  if(sum(colSums(data_temp$reads)==0)>0){print(paste0(folder_names[i],": empty motus present"))}
  if(sum(rowSums(data_temp$reads)==0)>0){print(paste0(folder_names[i],": empty pcrs present"))}
}
```

Remove all pcrs with no reads
```{r}
for (i in 1:length(folder_names)){
  data_temp <- metablists[[folder_names[i]]]
  data_temp <- subset_metabarlist(data_temp,"reads",rowSums(data_temp$reads)!=0)
  metablists[[folder_names[i]]] <- data_temp
} 
```

rm temp_files 
```{r}
toremove <- grep("folder_names|^metab", ls(), invert=T, value=T)
rm(list = c(toremove, "toremove"))
save.image("data/raw_metablists_jm_its.RData")
```