script to analyse data coming from my metabarcoding processing pipeline (OBIholo_pipeline).


# Input format


The basic data format used in `metabaR` is a `metabarlist`, a list of four tables:   

- `reads` a table of class `matrix` consisting of PCRs as rows, and molecular operational taxonomic units (MOTUs) as columns. The number of reads for each MOTU in each PCR is given in each cell, with 0 corresponding to no reads.   

- `motus` a table of class `data.frame` where MOTUs are listed as rows, and their attributes as columns. **A mandatory field in this table is the field "sequence", i.e. the DNA sequence representative of the MOTU**.  Examples of other attributes that can be included in this table are the MOTU taxonomic information, the taxonomic assignment scores, MOTU sequence GC content, MOTU total abundance in the dataset, etc. 

- `pcrs` a table of class `data.frame` consisting of PCRs as rows, and PCR attributes as columns. This table is particularly important in `metabaR`, as it contains all the information related to the extraction, PCR and sequencing protocols and design that are necessary to assess and improve the quality of metabarcoding data [@taberlet2018environmental; @zinger2019dna]. This table can also include information relating to the PCR design, such as the well coordinates and plate of each PCR, the tag combinations specific of the PCR, the primers used, etc. **Mandatory fields are**:    
    - **`sample_id`**: a vector indicating the biological sample origin of each PCR (e.g. the sample name)
    - **`type`** : the type of PCR, either a sample or an experimental control amplification. Only two values allowed: `"sample"` or `"control"`.     
    - **`control_type`** : the type of control. Only five values are possible in this field:    
        - `NA` if `type="sample"`, i.e. for any PCR obtained from a biological sample.    
        - `"extraction"` for DNA extraction negative controls, i.e. PCR amplification of an extraction where the DNA template was replaced by extraction buffer.   
        - `"pcr"` for PCR negative controls, i.e. pcr amplification where the DNA template was replaced by PCR buffer or sterile water.    
        - `"sequencing"` for sequencing negative controls, i.e. unused tag/library index combinations.     
        - `"positive"` for DNA extraction or PCR positive controls, i.e. pcr amplifications of known biological samples or DNA template (e.g. a mock community).     

- `samples` a table of class `data.frame` consisting of biological samples as rows, and associated information as columns. Such information includes e.g. geographic coordinates, abiotic parameters, experimental treatment, etc. This table does not include information on the DNA metabarcoding experimental controls, which can only be found in `pcrs`.   

# Prep files


Load required files:

```{r}
folder_names <- list.files("results")
list_motus <- list()
list_reads <- list()
list_filters <- list()
for (i in 1:length(folder_names)){
  current_name <- folder_names[i] 
  list_motus[[current_name]] <- read.csv2(paste0("../../OBIholo_test/results/",current_name,"/",current_name,"_taxassigned.csv"))
  list_reads[[current_name]] <- read_delim(paste0("../../OBIholo_test/results/",current_name,"/",current_name,"_R1R2_good_demultiplexed_basicfilt_derepl_cleaned_cl_agg.tab"),"\t", escape_double = F, trim_ws = T)
  list_filters[[current_name]] <- read_delim(paste0("../../OBIholo_test/resources/",current_name,"/",current_name,"_ngsfilter.tab"),"\t", escape_double = F, trim_ws = T)
} 
```


get MOTUS counts in data_motus
```{r}
for (i in 1:length(list_files)){
  list_motus[[i]] <- list_motus[[i]][match(toupper(list_reads[[i]]$sequence ),list_motus[[i]]$sequences ),]
  list_motus[[i]]$counts <- list_reads[[i]]$count
} 
```

prepare data_reads
```{r}
list_motus_seq <- list()
for (i in 1:length(list_reads)){
  list_motus_seq[[list_files[i]]] <-  list_reads[[i]][,c("id","sequence")] 
  sample_col <- which(grepl("sample",colnames(list_reads[[i]]))==T)
  tmp <- list_reads[[i]][,c(1,sample_col)]
  tmp1 <- t(tmp[,-1])
  id <- t(tmp[,1])
  colnames(tmp1) <- id
  rownames(tmp1) <- gsub("sample:","",rownames(tmp1))
  list_reads[[i]] <- tmp1 
} 
```

prepare data_samples
```{r}
list_samples <- list()
for (i in 1:length(list_reads)){
  list_samples[[i]] <- data.frame(samples=rownames(list_reads[[i]]))
} 
```

prepare data_pcrs
```{r}
list_pcrs <- list()
for (i in 1:length(list_reads)){
  colnames(list_filters[[i]]) <- c("project","ID","Tags","PrimeF","PrimeR","diff_doublons","x","Type")
  list_filters[[i]] <- dplyr::filter(list_filters[[i]], !ID %in% c(setdiff(sort(list_filters[[i]]$ID),sort(list_samples[[i]]$samples))))
  list_pcrs[[i]] <- list_samples[[i]] %>% mutate(type=ifelse(grepl("control",arrange(list_filters[[i]],ID)$Type)==T,"control","sample"),
                                                 control_type=ifelse(grepl("control_type=",arrange(list_filters[[i]],ID)$Type)==T,
                                                                     str_extract(arrange(list_filters[[i]],ID)$Type,"(?<=control_type=)\\w+"),NA),
                                                 plate= str_extract(arrange(list_filters[[i]],ID)$Type,"(?<=position=)[[:alnum:]]*"),
                                                 well= str_extract(arrange(list_filters[[i]],ID)$Type,"(?<=_)[[:upper:]]{1}[[:digit:]]{1,2}"))
}
```

check names issues
```{r}
for (i in 1:length(list_reads)){
  print(setdiff(list_samples[[i]]$samples,list_filters[[i]]$ID))
  print(setdiff(list_filters[[i]]$ID,list_samples[[i]]$samples))
} 
```

finalize
```{r}
for (i in 1:length(list_reads)){
  rownames(list_motus[[i]]) <- sprintf("MOTU_%05d", 1:nrow(list_motus[[i]]))
  colnames(list_reads[[i]]) <- sprintf("MOTU_%05d", 1:ncol(list_reads[[i]]))
  list_pcrs[[i]] %<>% arrange(samples)
  list_filters[[i]] %<>% arrange(ID)
  rownames(list_pcrs[[i]]) <- list_pcrs[[i]]$samples 
  colnames(list_motus[[i]])[1] <- "sequence" 
  
  list_pcrs[[i]] %<>% mutate(tag_fwd = str_extract(list_filters[[i]]$Tags,"^\\w+"),
                             tag_rev = str_extract(list_filters[[i]]$Tags,"\\w+$"),
                             primer_fwd = list_filters[[i]]$PrimeF,
                             primer_rev = list_filters[[i]]$PrimeR,
                             plate_col = as.numeric(str_extract(well,"[:digit:]{1,2}")),# numeric fÃ¼r metabaR 
                             plate_row = str_extract(well,"[:upper:]"),
                             plate= as.numeric(str_extract(plate,"[:digit:]{1,2}"))
  ) %>%
    rename(sample_id = samples,
           plate_no = plate) 
  
  list_pcrs[[i]] %<>%
    mutate(control_type = ifelse(control_type=="ctab","extraction",control_type))%>%
    mutate(control_type = as.factor(control_type))
  
  rownames(list_samples[[i]]) <- list_samples[[i]]$samples
}
```


generate metabarlists
```{r}
for (i in 1:length(list_reads)){
  assign(paste0("metabarlist_", list_files[i]),metabaR::metabarlist_generator(reads=as.matrix(list_reads[[i]]),
                                                                              motus=list_motus[[i]],
                                                                              pcrs=list_pcrs[[i]],
                                                                              samples=list_samples[[i]]))
}
```

Check if and which list have empty pcrs and motus.
```{r}
for (i in 1:length(list_files)){
  data_temp <- get(paste0("metabarlist_",list_files[i]))
  if(sum(colSums(data_temp$reads)==0)>0){print(paste0(list_files[i],": empty motus present"))}
  if(sum(rowSums(data_temp$reads)==0)>0){print(paste0(list_files[i],": empty pcrs present"))}
} 
```

Remmove all pcrs with no reads
```{r}
for (i in 1:length(list_files)){
  data_temp <- get(paste0("metabarlist_",list_files[i]))
  data_temp <- subset_metabarlist(data_temp,"reads",rowSums(data_temp$reads)!=0)
  assign(paste0("metabarlist_",list_files[i]),data_temp)
} 
```



# Add sample biological information


Add info on my samples, has to be improved but i'll see when i'll get my data.
```{r}
metabarlist_comp_mother_16s$samples <- metabarlist_comp_mother_16s$samples
metabarlist_comp_mother_16s$samples %<>% mutate(organ= ifelse(grepl("_F",metabarlist_comp_mother_16s$samples$samples)==T,"LEAVES",
                                                              ifelse(grepl("_R",metabarlist_comp_mother_16s$samples$samples)==T,"ROOTS",
                                                                     ifelse(grepl("_GR",metabarlist_comp_mother_16s$samples$samples)==T,"SEEDS","SOIL"))))

metabarlist_mother_16s$samples %<>% mutate(organ = ifelse( grepl("(_Rp)|([0-9]{3}R)",metabarlist_mother_16s$samples$samples)==T,"ROOTS",
                                                           ifelse( grepl("FL",metabarlist_mother_16s$samples$samples)==T,"FLOWER",
                                                                   ifelse( grepl("FE",metabarlist_mother_16s$samples$samples)==T,"LEAVES",
                                                                           ifelse( grepl("FR",metabarlist_mother_16s$samples$samples)==T,"FRUIT",
                                                                                   ifelse( grepl("RJ",metabarlist_mother_16s$samples$samples)==T,"RESHOOT",
                                                                                           ifelse( grepl("GR",metabarlist_mother_16s$samples$samples)==T,"SEEDS",
                                                                                                   ifelse( grepl("ED",metabarlist_mother_16s$samples$samples)==T,"ENDOPHYTES",
                                                                                                           ifelse( grepl("EP",metabarlist_mother_16s$samples$samples)==T,"EPIPHYTES",
                                                                                                                   ifelse( grepl("[0-9]{3}(S|C)",metabarlist_mother_16s$samples$samples)==T,"RHIZOSPHERE",
                                                                                                                           ifelse( grepl("CT|EXT|PCR|TPOS",metabarlist_mother_16s$samples$samples)==T,"CONTROL","SOIL")
                                                                                                                   ))))))))))

metabarlist_mother_glom$samples %<>% mutate(organ = ifelse( grepl("_R|(-R)|([0-9]{3}R)",metabarlist_mother_glom$samples$samples)==T,"ROOTS",
                                                            ifelse( grepl("FL",metabarlist_mother_glom$samples$samples)==T,"FLOWER",
                                                                    ifelse( grepl("FE",metabarlist_mother_glom$samples$samples)==T,"LEAVES",
                                                                            ifelse( grepl("FR",metabarlist_mother_glom$samples$samples)==T,"FRUIT",
                                                                                    ifelse( grepl("RJ",metabarlist_mother_glom$samples$samples)==T,"RESHOOT",
                                                                                            ifelse( grepl("GR",metabarlist_mother_glom$samples$samples)==T,"SEEDS",
                                                                                                    ifelse( grepl("ED",metabarlist_mother_glom$samples$samples)==T,"ENDOPHYTES",
                                                                                                            ifelse( grepl("EP",metabarlist_mother_glom$samples$samples)==T,"EPIPHYTES",
                                                                                                                    ifelse( grepl("[0-9]{3}(S|C)",metabarlist_mother_glom$samples$samples)==T,"RHIZOSPHERE",
                                                                                                                            ifelse( grepl("CT|EXT|PCR|TPOS",metabarlist_mother_glom$samples$samples)==T,"CONTROL","SOIL")
                                                                                                                    ))))))))))

metabarlist_seedling_16s$samples %<>% mutate(organ= ifelse(grepl("_F",metabarlist_seedling_16s$samples$samples)==T,"LEAVES",
                                                           ifelse(grepl("_R",metabarlist_seedling_16s$samples$samples)==T,"ROOTS","CONTROL")))


metabarlist_mother_ITS2$samples %<>% mutate(organ = ifelse( grepl("_R|(-R)|([0-9]{3}R)",metabarlist_mother_ITS2$samples$samples)==T,"ROOTS",
                                                            ifelse( grepl("FL",metabarlist_mother_ITS2$samples$samples)==T,"FLOWER",
                                                                    ifelse( grepl("FE",metabarlist_mother_ITS2$samples$samples)==T,"LEAVES",
                                                                            ifelse( grepl("FR",metabarlist_mother_ITS2$samples$samples)==T,"FRUIT",
                                                                                    ifelse( grepl("RJ",metabarlist_mother_ITS2$samples$samples)==T,"RESHOOT",
                                                                                            ifelse( grepl("GR",metabarlist_mother_ITS2$samples$samples)==T,"SEEDS",
                                                                                                    ifelse( grepl("ED",metabarlist_mother_ITS2$samples$samples)==T,"ENDOPHYTES",
                                                                                                            ifelse( grepl("EP",metabarlist_mother_ITS2$samples$samples)==T,"EPIPHYTES",
                                                                                                                    ifelse( grepl("[0-9]{3}(S|C)",metabarlist_mother_ITS2$samples$samples)==T,"RHIZOSPHERE",
                                                                                                                            ifelse( grepl("CT|EXT|PCR|TPOS",metabarlist_mother_ITS2$samples$samples)==T,"CONTROL","SOIL")
                                                                                                                    ))))))))))

metabarlist_seedling_ITS2$samples %<>% mutate(organ= ifelse(grepl("_F",metabarlist_seedling_ITS2$samples$samples)==T,"LEAVES",
                                                            ifelse(grepl("_R",metabarlist_seedling_ITS2$samples$samples)==T,"ROOTS","CONTROL")))


toremove <- grep("^metabarlist|list_files|list_motus_seq", ls(), invert=T, value=T)
rm(list = c(toremove, "toremove"))
```




# Metabar plot and diagnostic noise

Clean env so i just keep my metabarlists and filename list.
```{r}
toremove <- grep("^metabarlist|list_files|list_motus_seq", ls(), invert=T, value=T)
rm(list = c(toremove, "toremove"))
```


Plot the number of Reads and OTUs per pcr according to pcr type.
```{r}
diag_plots <- NULL
for (i in 1:length(list_files)){
  data_temp <- get(paste0("metabarlist_",list_files[[i]]))
  
  # Compute the number of reads per pcr
  data_temp$pcrs$nb_reads <- rowSums(data_temp$reads)
  
  # Compute the number of motus per pcr
  data_temp$pcrs$nb_motus <- rowSums(data_temp$reads>0)
  assign(paste0("metabarlist_",list_files[[i]]),data_temp)
  check1 <- melt(data_temp$pcrs[,c("control_type", "nb_reads", "nb_motus")])
  
  diag_plots[[list_files[i]]] <- ggplot(data <- check1, aes(x=control_type, y=value, color=control_type)) + 
    geom_boxplot() + theme_bw() + 
    geom_jitter(alpha=0.2) + 
    scale_color_manual(values = c("brown", "red", "cyan4","pink"), na.value = "darkgrey") +
    facet_wrap(~variable, scales = "free_y") + 
    theme(axis.text.x = element_text(angle=45, h=1))+
    ggtitle(list_files[[i]])
} 
diagplot <- ggarrange(plotlist = diag_plots, ncol = 2, nrow = length(diag_plots)/2, common.legend = T, legend = "bottom")
diagplot
```

Plot MOTU-reads correlation in each pcrs.
```{r}
motu_read_plots <- NULL 
for (i in 1:length(list_files)){
  # Using the nb_reads and nb_motus defined previously in the data_temp$pcrs table
  data_temp <- get(paste0("metabarlist_",list_files[[i]]))
  motu_read_plots[[list_files[i]]] <- ggplot(data_temp$pcrs, aes(x=nb_reads, y=nb_motus, color = control_type)) + 
    geom_point() + theme_bw() + 
    scale_y_log10() + scale_x_log10() + 
    scale_color_manual(values = c("brown", "red", "cyan4","pink"), na.value = "darkgrey")+
    ggtitle(list_files[[i]])
} 
moturead <- ggarrange(plotlist = motu_read_plots, ncol = 2, nrow = length(motu_read_plots)/2, common.legend = T, legend = "bottom")
moturead
```

rarefaction plots
```{r}
raref_plots <- NULL
for (i in 1:length(list_files)){
  data_temp <- get(paste0("metabarlist_",list_files[[i]]))
  raref <- hill_rarefaction(data_temp, nboot = 100, nsteps = 10)
  raref_plots[[list_files[i]]] <- gghill_rarefaction(raref) + ggtitle(list_files[[i]])
} 
raref <- ggarrange(plotlist =  raref_plots, ncol = 1, common.legend = T, legend ="bottom")
raref
```

identify extraction contaminants
```{r}
ext_conta_tables <- NULL
for (i in 1:length(list_files)){
  data_temp <- get(paste0("metabarlist_",list_files[[i]]))
  
# Identifying extraction contaminants
  data_temp <- contaslayer(data_temp, 
                          control_types = "extraction",
                          output_col = "not_an_extraction_conta")
  assign(paste0("metabarlist_",list_files[[i]]),data_temp)
  ext_conta_tables[[list_files[i]]] <- table(data_temp$motus$not_an_extraction_conta)
  print(ext_conta_tables[[i]])
  flush.console()
}
```

get their position in the PCR context
```{r}
conta_pos <- NULL
for (i in 1:length(list_files)){
  data_temp <- get(paste0("metabarlist_",list_files[[i]]))
  
# Identify the most common contaminant
# get contaminant ids
  id <- !data_temp$motus$not_an_extraction_conta
  max.conta <- rownames(data_temp$motus[id,])[which.max(data_temp$motus[id, "counts"])]
  if(length(max.conta)!=0){
  #... and its distribution and relative abundance in each pcr
  conta_pos[[list_files[i]]] <-  ggpcrplate(data_temp, legend_title = "#reads of most \nabundant contaminant",
             FUN = function(m) {m$reads[, max.conta]/rowSums(m$reads)})
  }
}
```

store and plot the proportion of contaminant reads per pcr and overall. 
```{r}
conta_prop <- NULL
conta_prop_table <- NULL
for (i in 1:length(list_files)){
  data_temp <- get(paste0("metabarlist_",list_files[[i]]))
  
  # Compute relative abundance of all pcr contaminants together 
  a <- data.frame(conta.relab = rowSums(data_temp$reads[,!data_temp$motus$not_an_extraction_conta]) / 
                    rowSums(data_temp$reads))
  # Add information on control types
  a$control_type <- data_temp$pcrs$control_type[match(rownames(a), rownames(data_temp$pcrs))]
  
  conta_prop[[list_files[i]]] <- ggplot(a, aes(x=control_type, y=conta.relab, color=control_type)) + 
    geom_boxplot() + geom_jitter(alpha=0.5) +
    scale_color_manual(values = c("brown", "red", "cyan4","pink"), na.value = "darkgrey") +
    labs(x=NULL, y="Prop. Reads (log10)") + 
    theme_bw() + 
    scale_y_log10() +
    ggtitle(list_files[[i]])
  
  data_temp$pcrs$low_contamination_level <- 
    ifelse(a$conta.relab[match(rownames(data_temp$pcrs), rownames(a))]>1e-1,  F, T)
  
  # Proportion of potentially functional (TRUE) vs. failed (FALSE) pcrs
  # (controls included) based on this criterion
  conta_prop_table[[list_files[i]]] <- table(data_temp$pcrs$low_contamination_level) / nrow(data_temp$pcrs)
  print(conta_prop_table[[i]])
  assign(paste0("metabarlist_",list_files[[i]]),data_temp)
  
}
contaprop <- ggarrange(plotlist =  conta_prop, ncol = 2,nrow = length(list_files)/2, common.legend = T , legend ="bottom")
contaprop
```

flag non-target motus.
```{r}
non_target_prop <- NULL
non_targetconta_prop <- NULL
for (i in 1:length(list_files)){
  
  data_temp <- get(paste0("metabarlist_",list_files[[i]]))
  target <- dplyr::last(BBmisc::explode(list_files[i],"_"))
  target_kingdom <- ifelse(target=="16s","Bacteria","Fungi")
  #Flag MOTUs corresponding to target (TRUE) vs. non-target (FALSE) taxa 
  data_temp$motus$target_taxon <- grepl(target_kingdom, data_temp$motus$Kingdom)
  
  # Proportion of each of these over total number of MOTUs
  non_target_prop[[list_files[i]]] <- table(data_temp$motus$target_taxon) / nrow(data_temp$motus)
  
  # Intersection with extraction contaminant flags (not contaminant = T)
  non_targetconta_prop[[list_files[i]]] <- table(data_temp$motus$target_taxon, 
                                     data_temp$motus$not_an_extraction_conta)
  print(non_target_prop[[i]]);print(non_targetconta_prop[[i]])
  assign(paste0("metabarlist_",list_files[[i]]),data_temp)
}
```

investigate poorly assignated sequence.
Need to set a confidence level or a taxonomic level were i reject either the assignation or even the sequence.
```{r}
plot_tax_boots <- NULL
for (i in 1:length(list_files)){
  
  data_temp <- get(paste0("metabarlist_",list_files[[i]]))
  melty <- melt(data_temp$motus, id = c(colnames(data_temp$motus)[which(grepl("_boots",colnames(data_temp$motus))==F)] ))
  melty$variable <- gsub("_boots","",melty$variable)
  melty$variable <- as.factor(melty$variable)
  melty$variable <- fct_relevel(melty$variable,"Kingdom","Phylum","Class","Order","Family","Genus","Species")
  plot_tax_boots[[list_files[i]]] <-  ggplot(melty,aes(x=variable,y=as.numeric(value))) +
                                   geom_boxplot(aes(fill=variable))+
    theme_bw()+
    ggtitle(list_files[[i]])+
    labs(y="conf %")+
    theme(panel.grid.major.x = element_blank(),
          panel.grid.minor.x = element_blank(),
          axis.title.x = element_blank())
}
plottaxboots <- ggarrange(plotlist =  plot_tax_boots, ncol = 2,nrow = length(list_files)/2, common.legend = T , legend ="bottom")
plottaxboots
```

check seq depth per pcrs
```{r}
plot_seq_depth <- NULL
seq_depth_tab <- NULL
for (i in 1:length(list_files)){

  data_temp <- get(paste0("metabarlist_",list_files[[i]]))
  plot_seq_depth[[list_files[i]]] <-  ggplot(data_temp$pcrs, aes(nb_reads)) +
    geom_histogram(bins=40, color="grey", fill="white") + 
    geom_vline(xintercept = 1e3, lty=2, color="orange") + # threshold
    scale_x_log10() + 
    labs(x="# Reads (with all MOTUs and PCRs)", 
         y="# PCRs") +
    theme_bw() + 
    theme(panel.grid = element_blank())+
    ggtitle(list_files[[i]])
  
  
  data_temp$pcrs$seqdepth_ok <- ifelse(data_temp$pcrs$nb_reads > 1e3, F, T)
  assign(paste0("metabarlist_",list_files[[i]]),data_temp)
  
  seq_depth_tab[[list_files[i]]] <- table(data_temp$pcrs$seqdepth_ok[data_temp$pcrs$type=="sample"]) /
    nrow(data_temp$pcrs[data_temp$pcrs$type=="sample",])
  print(seq_depth_tab[[i]])
}
plotseqdepth <- ggarrange(plotlist =  plot_seq_depth, ncol = 2,nrow = length(list_files)/2)
plotseqdepth
```


Test different thresholds to identify tagjumps 
```{r}
plot_tag_jump <- NULL
for (i in 1:length(list_files)){
  
  
  data_temp <- get(paste0("metabarlist_",list_files[[i]]))
  
  # Define a vector of thresholds to test
  thresholds <- c(0,1e-4,1e-3, 1e-2, 3e-2, 5e-2) 
  
  # Run the tests and stores the results in a list
  tests <- lapply(thresholds, function(x) tagjumpslayer(data_temp,x))
  names(tests) <- paste("t_", thresholds, sep="")
  
  # Format the data for ggplot with amount of reads at each threshold
  tmp <- melt(as.matrix(do.call("rbind", lapply(tests, function(x) rowSums(x$reads)))))
  colnames(tmp) <- c("threshold", "sample", "abundance")
  
  # Add richness in MOTUs at each threshold
  tmp$richness <-
    melt(as.matrix(do.call("rbind", lapply(tests, function(x) {
      rowSums(x$reads > 0)
    }))))$value
  
  # Add control type information on pcrs and make data curation threshold numeric
  tmp$controls <- data_temp$pcrs$control_type[match(tmp$sample, rownames(data_temp$pcrs))]
  tmp$threshold <- as.numeric(gsub("t_", "", tmp$threshold))
  
  # New table formatting for ggplot
  tmp2 <- melt(tmp, id.vars=colnames(tmp)[-grep("abundance|richness", colnames(tmp))])
  
  plot_tag_jump[[list_files[i]]] <- ggplot(tmp2, aes(x=as.factor(threshold), y=value)) + 
    geom_boxplot(color="grey40") + 
    geom_vline(xintercept = which(levels(as.factor(tmp2$threshold)) == "0.01"), col="orange", lty=2) + 
    geom_jitter(aes(color=controls), width = 0.2, alpha=0.5) + 
    scale_color_manual(values = c("brown", "red", "cyan4","pink"), na.value = "darkgrey") +
    facet_wrap(~variable+controls, scale="free_y", ncol=5) + 
    theme_bw() + 
    scale_y_log10() +
    labs(x="MOTU pcr : total abundance filtering threshold", y="# Reads/MOTUs") + 
    theme(panel.grid = element_blank(), 
          strip.background = element_blank(), 
          axis.text.x = element_text(angle=40, h=1), 
          legend.position = "none")+
    ggtitle(list_files[[i]])
}

plottagjump <- ggarrange(plotlist =  plot_tag_jump, ncol = 2,nrow = length(list_files)/2)
plottagjump
```

summarize noise in motus
```{r}
plot_noise_motu <- NULL
motu_noise_tab <- NULL
for (i in 1:length(list_files)){
  
  data_temp <- get(paste0("metabarlist_",list_files[[i]]))
  
  # Create a table of MOTUs quality criteria 
  # noise is identified as FALSE in data_temp, the "!" transforms it to TRUE 
  motus.qual <- !data_temp$motus[,c("not_an_extraction_conta", "target_taxon")] #not_degraded 
  colnames(motus.qual) <- c("extraction_conta", "untargeted_taxon") #degraded_seq  
  
  # Proportion of MOTUs potentially artifactual (TRUE) based on the criteria used
  motu_noise_tab[[list_files[i]]]$motus <- prop.table(table(apply(motus.qual, 1, sum) > 0))
  
  # Corresponding proportion of artifactual reads (TRUE)
  motu_noise_tab[[list_files[i]]]$reads <- prop.table(xtabs(data_temp$motus$count~apply(motus.qual, 1, sum) > 0))
  
  # Proportion of MOTUs and reads potentially artifactual for each criterion
  apply(motus.qual, 2, sum) / nrow(motus.qual)
  apply(motus.qual, 2, function(x) sum(data_temp$motus$count[x])/sum(data_temp$motus$count))
  
  tmp.motus <- 
    apply(sapply(1:ncol(motus.qual), function(x) {
      ifelse(motus.qual[,x]==T, colnames(motus.qual)[x], NA)}), 1, function(x) {
        paste(sort(unique(x)), collapse = "|")
      })
  tmp.motus <- as.data.frame(gsub("^$", "not_artefactual", tmp.motus))
  colnames(tmp.motus) <-  "artefact_type"
  
  plot_noise_motu[[list_files[i]]] <- ggplot(tmp.motus, aes(x=1, fill=artefact_type)) +
    geom_bar() +  xlim(0, 2) +
    labs(fill="Artifact type") + 
    coord_polar(theta="y") + theme_void() + 
    scale_fill_brewer(palette = "Set3") + 
    theme(legend.direction = "vertical") + 
    ggtitle(list_files[i])
}
plotnoisemotu <- ggarrange(plotlist =  plot_noise_motu, ncol = 2,nrow = length(list_files)/2,common.legend = T, legend = "bottom")
plotnoisemotu
```

summarize noise in pcrs
```{r}
plot_noise_pcr <- NULL
noise_pcr_tab <- NULL
for (i in 1:length(list_files)){
  
  data_temp <- get(paste0("metabarlist_",list_files[[i]]))
  
  # Create a table of pcrs quality criteria 
  # noise is identified as FALSE in data_temp, the "!" transforms it to TRUE
  pcrs.qual <- !data_temp$pcrs[,c("low_contamination_level", "seqdepth_ok")] #  "replicating_pcr"
  colnames(pcrs.qual) <- c("high_contamination_level", "low_seqdepth") # "outliers"
  
  # Proportion of pcrs potentially artifactual (TRUE) based on the criteria used
  # excluding controls
  noise_pcr_tab[[list_files[i]]] <- prop.table(table(apply(pcrs.qual[data_temp$pcrs$type=="sample",], 1, sum) > 0))
  
  # Proportion of MOTUs and reads potentially artifactual for each criterion
  apply(pcrs.qual[data_temp$pcrs$type=="sample",], 2, sum) / nrow(pcrs.qual[data_temp$pcrs$type=="sample",])
  
  tmp.pcrs <- 
    apply(sapply(1:ncol(pcrs.qual), function(x) {
      ifelse(pcrs.qual[data_temp$pcrs$type=="sample",x]==T, 
             colnames(pcrs.qual)[x], NA)}), 1, function(x) {
               paste(sort(unique(x)), collapse = "|")
             })
  tmp.pcrs <- as.data.frame(gsub("^$", "not_artefactual", tmp.pcrs))
  
  colnames(tmp.pcrs) <- "artefact_type"
  
  plot_noise_pcr[[list_files[i]]] <- ggplot(tmp.pcrs, aes(x=1, fill=artefact_type)) +
    geom_bar() +  xlim(0, 2) +
    labs(fill="Artifact type") + 
    coord_polar(theta="y") + theme_void() + 
    scale_fill_brewer(palette = "Set3") + 
    theme(legend.direction = "vertical") +
    ggtitle(list_files[i])
  
} 
  
plotnoisepcr <- ggarrange(plotlist =  plot_noise_pcr, ncol = 2,nrow = length(list_files)/2,common.legend = T, legend = "bottom")
plotnoisepcr
```






# Clean noise

```{r}
cleaned_metablist <- NULL
summary_list <- NULL
sumpipeline_plots <- NULL
for (i in 1:length(list_files)){ 
  # Use tag-jump corrected metabarlist with the threshold identified above
  tmp <- test_metablist[[i]][["t_0.01"]]
  
  # Subset on MOTUs: we keep motus that are defined as TRUE following the 
  # three criteria below (sum of three TRUE is equal to 3 with the rowSums function)
  tmp <- subset_metabarlist(tmp, "motus", 
                            indices = rowSums(tmp$motus[,c("not_an_extraction_conta", "target_taxon")]) == 2)#not_degraded 
  
  # Subset on pcrs and keep only controls 
  cleaned_metablist[[list_files[i]]] <- data_temp <- subset_metabarlist(tmp, "pcrs", 
                                                                        indices = rowSums(tmp$pcrs[,c("low_contamination_level", "seqdepth_ok")]) == 2 & #, "replicating_pcr" 
                                                                          tmp$pcrs$type == "sample")
  summary_list[[list_files[i]]] <- summary_metabarlist(data_temp)
  
  if(sum(colSums(data_temp$reads)==0)>0){print(paste0(list_files[[i]]," empty motus present"))}
  if(sum(colSums(data_temp$reads)==0)>0){print(paste0(list_files[[i]]," empty pcrs present"))}
  
  #update counts and reads.    
  data_temp$motus$count = colSums(data_temp$reads)
  data_temp$pcrs$nb_reads_postmetabaR = rowSums(data_temp$reads)
  data_temp$pcrs$nb_motus_postmetabaR = rowSums(ifelse(data_temp$reads>0, T, F))
  
  check <- melt(data_temp$pcrs[,c("reads", "reads_postmetabaR", 
                                       "motus", "motus_postmetabaR")])
  check$type <- ifelse(grepl("motus", check$variable), "richness", "abundance")
  
  sumpipeline_plots[[list_files[i]]] <- ggplot(data = check, aes(x = variable, y = value)) +
    geom_boxplot( color = "darkgrey") +
    geom_jitter(alpha=0.1, color = "darkgrey") +
    theme_bw() +
    facet_wrap(~type, scales = "free", ncol = 5) +
    theme(axis.text.x = element_text(angle=45, h=1),
          axis.title = element_blank()) +
    ggtitle(list_files[i])
  
  cleaned_metablist[[list_files[i]]] <- data_temp
}  
sumpipelineplots <- ggarrange(plotlist =  sumpipeline_plots, ncol = 2,nrow = length(list_files)/2,common.legend = T, legend = "bottom")
sumpipelineplots
```

Visualize the effect of the cleaning on the biological signal.
```{r}

beta_div <- NULL
for (i in 1:length(list_files)){ 
  data_temp_cleaned <- cleaned_metablist[[i]] 
  data_temp <- get(paste0("metabarlist_",list_files[i]))
  # Get row data only for samples
  tmp <- subset_metabarlist(data_temp, table = "pcrs",
                            indices = data_temp$pcrs$type == "sample")
  
  # Build PCoA ordinations 
  mds1 <- check_pcr_repl(tmp,
                         groups =  as.factor(tmp$samples$organ))
  mds2 <- check_pcr_repl(data_temp_cleaned,
                         groups =  as.factor(data_temp_cleaned$samples$organ))
  
  # Custom colors
  a <- mds1 + labs(color = "organ") +
    theme(legend.position = "none") + 
    ggtitle("Raw data")
  b <- mds2 + labs(color = "organ") +
    ggtitle("Clean data")
  
  # Assemble plots
  leg <- get_legend(b + guides(shape=F) + 
                      theme(legend.position = "right", 
                            legend.direction = "vertical"))
  beta_div[[list_files[i]]] <-  ggdraw() +
    draw_plot(a, x=0, y=0, width = 0.4, height = 1) + 
    draw_plot(b + guides(color=F, shape=F), x=0.42, y=0, width = 0.4, height = 1) +
    draw_grob(leg, x=0.4, y=0) +
    ggtitle(list_files[[i]])
  
}
betadiv <- ggarrange(plotlist =  beta_div, ncol = 2,nrow = length(list_files)/2,common.legend = T, legend = "bottom")
betadiv
```


